review: conservative_discrete_laplacian_tail_to_alpha

Shoeboxam opened the pull request:
Close #2061

In the "Files changed" diffs I added a space to the end of each lines to make it possible to leave comments inline. Either focus on leaving comments in the green, or comment on the pdf rendered below and upload it in your review. Thanks!

Shoeboxam committed:
review: conservative_discrete_laplacian_tail_to_alpha

Shoeboxam commented:
* **#2124** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2124?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

annabelle400 reviewed (commented):
In the proof of the theorem, isn't p = 1/e^{1/s} ?

The statement of the theorem doesn't seem to match the last line of the proof?

I think with the above value for p, the proof does yield the last line.

Can you say what scale and tail are in the theorem? I assume s, t ?

I am not sure now whether the code matches the last line of the proof, or the statement of the theorem on which it depends. Please check and clarify.

tudorcebere reviewed (changes requested):
The proof looks great! I would be tempted to make it more similar to the Gaussian one, where there are some explanations provided and a bit of context. I am happy to give it a try to add those. I also left a few comments (the floating points one might be wrong)

Shoeboxam committed:
respond to review 4/20

Shoeboxam commented:
Thanks for all the comments. I fixed the proof and normalized it to the newer template style from #2699. 

* corrected the pmf normalization sign
* corrected the theorem statement for P[X > t]
* clarified that s = scale and t = tail
* fixed the geometric-series substitution to use p = e^{-1/s}

On the rounding question: Tudor’s concern was valid for the pseudocode, but not for the Rust implementation itself. The implementation was already using the right conservative directions; the pseudocode was too indirect and made the numerator look downward-rounded. I rewrote it to directly mirror the implementation, as it should be.

I think this was likely a partial-update from an earlier iteration: parts of the proof seem to match an older version of the function. This would explain why the pseudocode no longer mirrored how the implementation actually rounded.

annabelle400 commented:
Dear Michael,

Thanks for these! I am travelling until early May. Would it be ok to look at these when I get back?

Thanks!
Annabelle

On 22 Apr 2026, at 4:58 am, Michael Shoemate ***@***.***> wrote:

﻿

@Shoeboxam<https://github.com/Shoeboxam> requested your review on: opendp/opendp#2124<https://github.com/opendp/opendp/pull/2124> review: conservative_discrete_laplacian_tail_to_alpha.

—
Reply to this email directly, view it on GitHub<https://github.com/opendp/opendp/pull/2124?email_source=notifications&email_token=ANUEODRKNBUADAUIAKWYGJ34XAYN3A5CNFSNUABQM5UWIORPF5TWS5BNNB2WEL2JONZXKZKFOZSW45CON52GSZTJMNQXI2LPNYXTENBXGQYDCOJZGM3TLJTSMVQXG33OWBZGK5TJMV3V64TFOF2WK43UMVSKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#event-24740199375>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANUEODTTPDTKLR75JTKSSJD4XAYN3AVCNFSM6AAAAACV5I75JOVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMRUG42DAMJZHEZTONI>.
You are receiving this because your review was requested.Message ID: ***@***.***>

Shoeboxam commented:
@annabelle400 Of course. Safe travels!

annabelle400 reviewed (approved):
Looks ok now.
